sj201_interface/util/pwmcontroller.py

```diff
@@ -0,0 +1,144 @@
+        self.pwm1_range = ctypes.c_int.from_buffer(self._pwm_mv, PWMController.PWM1_RANGE*4)
+
+
+    def configure_pwm0(self, range: int):
+        self.pwm0_range.value = range
+        # print(f"Set PWM0 range to: {range}")
+
+    def configure_pwm1(self, range: int):
+        self.pwm1_range.value = range
+        # print(f"Set PWM1 range to: {range}")
+
+    def set_mode(self, channel_mode: str):
+        mode = PWMController.PWM0_ENABLE | PWMController.PWM1_ENABLE
+        if channel_mode == "ms":
+            mode = mode | PWMController.PWM0_MS_MODE | PWMController.PWM1_MS_MODE
+
+        self.pwm_control.value = mode
+        # print(f"Set PWM_CONTROL to: 0x{mode:08X}")
+
+    def pwm_set_clock(self, divisor: int):
+        divisor = int(540*divisor/192)
+        divisor &= 4095
+
+        pwm_clk_div = self.pwm_clk_div
+        pwm_control = self.pwm_control
+        
+        # print(f"Setting to: {divisor}. Current: 0x{pwm_clk_div.value:08X}")
+
+        with AutoRestoreRegisters([pwm_control]):
+            # print(f"Current PWM_CONTROL: {prev_pwm_control}")
+
+            # We need to stop PWM prior to stopping PWM clock in MS mode otherwise BUSY
+            # stays high.
+            pwm_control.value = 0 # Stop PWM
+
+            # Stop PWM clock before changing divisor. The delay after this does need to
+            # this big (95uS occasionally fails, 100uS OK), it's almost as though the BUSY
+            # flag is not working properly in balanced mode. Without the delay when DIV is
+            # adjusted the clock sometimes switches to very slow, once slow further DIV
+            # adjustments do nothing and it's difficult to get out of this mode.
+
+            pwm_clk_cntl = self.pwm_clk_cntl
+            pwm_clk_cntl.value = PWMController.BCM_PASSWORD | 0x01 # Stop PWM Clock
+            time.sleep(0.0001)                                  # prevents clock going sloooow
+
+            while ((pwm_clk_cntl.value & 0x80) != 0): # Wait for clock to be !BUSY
+                time.sleep(0.000001)
+
+            pwm_clk_div.value = PWMController.BCM_PASSWORD | (divisor << 12)
+            pwm_clk_cntl.value = PWMController.BCM_PASSWORD | 0x011 # Start PWM Clock
+
+        # print(f"Set     to: {divisor}. Now    : 0x{pwm_clk_div.value:08X}")
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def close(self):
+        del self.pwm0_range
+        del self.pwm1_range
+        del self.pwm_control
+        del self._pwm_mv
+        self._pwm.close()
+
+        del self.pwm_clk_div
+        del self.pwm_clk_cntl
+        del self._clk_mv
+        self._clk.close()
+
+if __name__ == "__main__":
+    pwm_controller = PWMController()
+    pwm_controller.set_mode("ms")
+    pwm_controller.configure_pwm1(32)
+    pwm_controller.pwm_set_clock(8)
+    pwm_controller.close()
```

Comment:
Is this used for something, or just leftover from testing? If it belongs in a CLI entrypoint, it can be added to `cli.py` and otherwise probably removed

Reply:
Addressed in 25d85623